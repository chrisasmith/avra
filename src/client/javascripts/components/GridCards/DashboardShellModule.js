import React, { Component } from 'react'
import classnames from 'classnames'
import PropTypes from 'prop-types'
import styles from './styles/DashboardModule.scss'
import TooltipPopup from '../TooltipPopup'

const propTypes = {
  title: PropTypes.string.isRequired,
  isFetching: PropTypes.bool.isRequired,
  children: PropTypes.element,

  onDownloadImage: PropTypes.func,
  onDownloadCSV: PropTypes.func,
}

class DashboardShellModule extends Component {
  constructor(props, context) {
    super(props, context)

    this.downloadButton = React.createRef()
    this.popup = React.createRef()

    this.onDownloadImage = this.onDownloadImage.bind(this)
    this.onDownloadClick = this.onDownloadClick.bind(this)
    this.closePopup = this.closePopup.bind(this)
    this.onDownloadCSV = this.onDownloadCSV.bind(this)

    this.state = {
      popupTarget: null,
      showDownloadPopup: false,
    }
  }

  openPopup() {
    this.setState({ showDownloadPopup: true })
  }

  closePopup() {
    this.setState({ showDownloadPopup: false })
  }

  onDownloadClick(e) {
    this.setState({ popupTarget: e.target }, () => {
      this.openPopup()
    })
  }

  onDownloadImage(e) {
    if (e) e.preventDefault()

    this.props.onDownloadImage()
    this.closePopup()
  }

  onDownloadCSV(e) {
    if (e) e.preventDefault()

    this.props.onDownloadCSV()
    this.closePopup()
  }

  render() {
    const {
      title,
      isFetching,
      children,
      onDownloadCSV,
      onDownloadImage,
    } = this.props


    if (isFetching) {
      return (
        <div className={styles.module}>
          <div className="container">
            <div>Loading...</div>
          </div>
        </div>
      )
    }

    return (
      <div className={styles.module}>
        <div className="container">
          <div
            href="#"
            className={styles.download}
            onClick={this.onDownloadClick}
          >
            <i className="fa fa-download" aria-hidden="true" />
          </div>
          <TooltipPopup
            target={this.state.popupTarget}
            show={this.state.showDownloadPopup}
            onClickOutside={this.closePopup}
          >
            <div className={styles.popup}>
              {onDownloadImage && (
                <a href="#" onClick={this.onDownloadImage}>Download .jpg</a>
              )}
              {onDownloadCSV && (
                <a href="#" onClick={this.onDownloadCSV}>Download csv</a>
              )}
            </div>
          </TooltipPopup>
          <h6 className={classnames('text-center pt-3 pb-2', styles.title)}>{title}</h6>
          {children}
        </div>
      </div>
    )
  }
}

DashboardShellModule.propTypes = propTypes
export default DashboardShellModule
