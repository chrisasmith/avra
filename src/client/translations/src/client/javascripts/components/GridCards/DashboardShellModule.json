[
  {
    "id": "dashboard_shell.loading",
    "defaultMessage": "Loading..."
  },
  {
    "id": "dashboard_shell.download_jpg",
    "defaultMessage": "Download .jpg"
  },
  {
    "id": "dashboard_shell.download_csv",
    "defaultMessage": "Download csv"
  }
]