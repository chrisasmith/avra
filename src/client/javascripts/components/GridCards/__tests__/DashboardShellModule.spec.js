import React from 'react'
import { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import DashboardShellModule from '../DashboardShellModule'

describe('DashboardShellModule Component', () => {
  it('Should show', () => {
    const wrapper = shallow(<DashboardShellModule
      title="title"
      isFetching={false}
    ><div>hi</div></DashboardShellModule>)

    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })

  it('Should show loading', () => {
    const wrapper = shallow(<DashboardShellModule
      title="title"
      isFetching
    ><div>hi</div></DashboardShellModule>)

    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })

  it('Should show wire up download buttons', () => {
    const wrapper = shallow(<DashboardShellModule
      title="title"
      isFetching={false}
      onDownloadImage={jest.fn()}
      onDownloadCSV={jest.fn()}
    ><div>hi</div></DashboardShellModule>)


    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })

  it('Should should fire download image event', () => {
    const onDownload = jest.fn()
    const wrapper = shallow(<DashboardShellModule
      title="title"
      isFetching={false}
      onDownloadImage={onDownload}
      onDownloadCSV={jest.fn()}
    ><div>hi</div></DashboardShellModule>)

    wrapper.find('a').at(0).simulate('click')
    expect(onDownload).toHaveBeenCalled()
  })

  it('Should should fire download csv event', () => {
    const onDownload = jest.fn()
    const wrapper = shallow(<DashboardShellModule
      title="title"
      isFetching={false}
      onDownloadImage={jest.fn()}
      onDownloadCSV={onDownload}
    ><div>hi</div></DashboardShellModule>)

    wrapper.find('a').at(1).simulate('click')
    expect(onDownload).toHaveBeenCalled()
  })
})
