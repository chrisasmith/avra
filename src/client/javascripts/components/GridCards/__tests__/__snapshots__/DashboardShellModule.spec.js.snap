// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardShellModule Component Should show 1`] = `
<div
  className="module"
>
  <div
    className="container"
  >
    <div
      className="download"
      href="#"
      onClick={[Function]}
    >
      <i
        aria-hidden="true"
        className="fa fa-download"
      />
    </div>
    <TooltipPopup
      onClickOutside={[Function]}
      show={false}
      target={null}
    >
      <div
        className="popup"
      />
    </TooltipPopup>
    <h6
      className="text-center pt-3 pb-2 title"
    >
      title
    </h6>
    <div>
      hi
    </div>
  </div>
</div>
`;

exports[`DashboardShellModule Component Should show loading 1`] = `
<div
  className="module"
>
  <div
    className="container"
  >
    <div>
      Loading...
    </div>
  </div>
</div>
`;

exports[`DashboardShellModule Component Should show wire up download buttons 1`] = `
<div
  className="module"
>
  <div
    className="container"
  >
    <div
      className="download"
      href="#"
      onClick={[Function]}
    >
      <i
        aria-hidden="true"
        className="fa fa-download"
      />
    </div>
    <TooltipPopup
      onClickOutside={[Function]}
      show={false}
      target={null}
    >
      <div
        className="popup"
      >
        <a
          href="#"
          onClick={[Function]}
        >
          Download .jpg
        </a>
        <a
          href="#"
          onClick={[Function]}
        >
          Download csv
        </a>
      </div>
    </TooltipPopup>
    <h6
      className="text-center pt-3 pb-2 title"
    >
      title
    </h6>
    <div>
      hi
    </div>
  </div>
</div>
`;
